File: 101-symmetric-tree/101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool isSymmetricHelper(TreeNode* left, TreeNode* right) {
        if(left == NULL || right == NULL) return left == right;
        if(left->val != right->val) return false;
        return isSymmetricHelper(left->left, right->right) && isSymmetricHelper(left->right, right->left);
    }
public:
    // TC -> O(n) SC -> O(n)
    bool isSymmetric(TreeNode* root) {
        return root == NULL || isSymmetricHelper(root->left, root->right);
    }
};

File: 101-symmetric-tree/101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
        if not left or not right: return left == right
        if left.val != right.val: return False
        return self.helper(left.left, right.right) and self.helper(left.right, right.left)
    
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return not root or self.helper(root.left, root.right)